contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+
+    function _deposit(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    ) private {
+        // assert valid amount sent
+        if (amount < minDepositAmount) {
+            revert TooSmallDeposit();
+        }
+
+        // transfer deposit token from portfolio
+        depositTokenInfo.token.safeTransferFrom(
+            _msgSender(),
+            address(this),
+            amount
+        );
+
+        // compute actual deposit and transfer fee to receiver
+        amount = _takeFee(amount);
+
+        DcaDepositor storage depositor = depositors[sender];
+
+        // assert positions limit is not reached
+        if (depositor.positions.length == positionsLimit) {
+            revert PositionsLimitReached();
+        }
+
+        // add splitted amounts to the queue
+        globalInvestQueue.splitUserInvestmentAmount(amount, amountSplit);
+
+        // if not started position with the same split exists - increase deposit amount
+        for (uint256 i = 0; i < depositor.positions.length; i++) {
+            // calculate amount of passed investment epochs
+            uint256 passedInvestPeriods = (lastInvestmentTimestamp -
+                depositor.positions[i].investedAt) / investmentPeriod;
+
+            if (
+                passedInvestPeriods == 0 &&
+                depositor.positions[i].amountSplit == amountSplit
+            ) {
+                // not started position with the same amount split exists
+                // just add invested amount here
+                depositor.positions[i].depositAmount += amount;
+                return;
+            }
+        }
+
+        // otherwise create new position
+        depositor.positions.push(
+            Position(
+                amount,
+                amountSplit,
+                lastInvestmentTimestamp,
+                dcaHistory.currentHistoricalIndex()
+            )
+        );
+
+        emit Deposit(sender, amount, amountSplit);
+    }
+
+    function invest()
+        public
+        virtual
+        onlyDcaInvestor
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        // declare total amount for event data
+        uint256 totalDepositSpent;
+        uint256 totalBluechipReceived;
+
+        // assert triggered at valid period
+        uint256 passedInvestPeriods = _getPassedInvestPeriods();
+        if (passedInvestPeriods == 0) {
+            revert NothingToInvest();
+        }
+
+        // iterate over passed invest periods
+        for (uint256 i = 0; i < passedInvestPeriods; i++) {
+            uint256 depositedAmount = globalInvestQueue
+                .getCurrentInvestmentAmountAndMoveNext();
+
+            // nobody invested in the queue, just skip this period
+            if (depositedAmount == 0) {
+                continue;
+            }
+
+            // swap deposit amount into invest token
+            uint256 receivedBluechip = _swapIntoBluechipAsset(depositedAmount);
```

Comment:
depending on how many times this loop executes (meaning how often we call the invest function), it might worth putting the swapping and investing outside the loop.

if the invest off-chain component is down for a long time, theoretically we could run out of block gas limit depending on how much gas the _invest requires